### ISSUE TITLE
Make it quick and easy to open a "file" referenced from a BibTeX entry

### ISSUE THREAD

**Issue by salutis** [bdarcus/bibtex-actions#191]
From https://github.com/tmalsburg/helm-bibtex/issues/383:

>> I added `file = {monet-1877a.jpg}` to my BibTeX entry. Then, I ran `bibtex-actions-open`, and it showed `[multiple] References: has:link\|has:pdf`. I wondered why the prompt is not`has:url\|has:file` instead, as those are the relevant BibTeX fields. (Right?)
> 
> So this piece is my problem :-)

I am not even sure what to ask for, but here we go.

Brainstorming off the top of my head:

I would expect `bibtex-actions-open-link` to be called `bibtex-actions-open-url`, if that is what the function does, and I would expect `bibtex-actions-open-file` to exist.

Then, the big question is, what shoud `bibtex-actions-open` do?

First and foremost, I would expect it to be intelligent. If the BibTeX entry has one `file` and nothing else (no notes, no URL, no nothing), then it should call `bibtex-actions-open-file`, as there is nothing else to open. If the BibTeX entry has more going on, I would expect Emacs to ask me.

Here is a basic 101 use-case:

1. `M-x bibtex-actions-open`
2. select some item and press `RET`
3. see the item (URL, file, ...)

I think the basic 101 use-case should "just work".

**Comment by bdarcus** [comment]
> I would expect `bibtex-actions-open-link` to be called `bibtex-actions-open-url`, if that is what the function does, and I would expect `bibtex-actions-open-file` to exist.

If you search the issue tracker and PRs, you'll find this isn't a new question. But it's not super straightforward.

Here's the current situation:

- Open-pdf = open local (PDF) file
- Open-url = open remote file/resource
- Open-doi = same
- Ooen-link = url OR doi (bibtex-completion uses the latter language, but I saw no need for that distinction)
- Open = open any of the above, in order (if PDF, do that, else ...)

These commands are just wrappers around bibtex-completion functions. I tried to name them to simplify, but in an earlier issue we concluded ideally "open" is really "open source."

But the specific command and function name that I think is at issue is really the first one; "pdf" is in the very function and command names. You really want that to be "open-local-file" (keep in mind there's always a tension between brevity and clarity though).

> Then, the big question is, what shoud `bibtex-actions-open` do?
> 
> First and foremost, I would expect it to be intelligent. If the BibTeX entry has one `file` and nothing else (no notes, no URL, no nothing), then it should call `bibtex-actions-open-file`, as there is nothing else to open. If the BibTeX entry has more going on, I would expect Emacs to ask me.

Per above, that's what it does.

I'll need to think more about what's to be done about the issue you raise, and where. More later.

**Comment by salutis** [comment]
> But the specific command and function name that I think is at issue is really the first one; "pdf" is in the very function and command names. You really want that to be "open-local-file" (keep in mind there's always a tension between brevity and clarity though).

True. Further, it also ties to `has:pdf` in the `bibtex-actions-open`. How about `bibtex-actions-open-file` and `has:file` instead? I am not sure if "local" is necessary; the term _file_ in its original sense is a perfect fit: "file (n.) a folder or box for holding loose papers that are typically arranged in a particular order for easy reference" (Oxford Dictionary, the primary definition). Do you think that bare "file" could be confused with some other file?

**Comment by salutis** [comment]
By the way, not to hijack this issue, but when I confirm `has:link\|has:pdf` in `actions-open-file`, it goes into "Selected". Is that normal?

<img width="947" alt="selection" src="https://user-images.githubusercontent.com/1158274/126916566-cdb5ed17-a8f0-432a-8754-33951f0c785d.png">

**Comment by bdarcus** [comment]
> By the way, not to hijack this issue, but when I confirm `has:link\|has:pdf` with `RET` in `actions-open-file`, it goes into "Selected". Is that normal?

Yes, IIRC it's (maybe necessary, or maybe temporary) limitation in `consult-completing-read-multiple`, which is very new.

But we probably need to ditch our method of pre-filtering candidates, as you see there. It's not really designed for this.

The problem is the "correct" solution would mean the candidates would be filtered, without ability to remove the filter.

**Comment by salutis** [comment]
@bdarcus 

> The problem is the "correct" solution would mean the candidates would be filtered, without ability to remove the filter.

That makes perfect sense.

**Comment by salutis** [comment]
To clarify:

Perhaps I am using `vertico` and `bibtex-actions` incorrectly? (I am new to both.) To make my situation obvious, here is what I am doing, step by step. First, I exectue `bibtex-actions-open` which I bound to `C-c b`, and Emacs immediately prompts me as follows:

```
*/0 [multiple] References: has:link\|has:pdf
```

At this point, if I start typing, no items show "live", like elsewhere in `vertico`/`consult`. If I type `monet`, Emacs puts `has:link\|has:pdf monet` item to the _Selected_ section. If I press `RET` on that item, I get `Wrong type argument: stringp, nil`. If, instead of typing `monet` at the beginning, I simply press `RET`, I get `has:link\|has:pdf` in the _Selected_ section.

In other words, `bibtex-actions-open` works "normally" only if I remove `has:link\|has:pdf`. But then, why would it be there? The fact that I am asking that question makes me wonder if perhaps I am missing some basic `vertico` concept, or perhaps my Emacs configuration is off?

**Comment by salutis** [comment]
> > But the specific command and function name that I think is at issue is really the first one; "pdf" is in the very function and command names. You really want that to be "open-local-file" (keep in mind there's always a tension between brevity and clarity though).

Or `-library-file`. These are the files in `bibtex-completion-library-path`, right?

**Comment by bdarcus** [comment]
> At this point, if I start typing, no items show "live", like elsewhere in `vertico`/`consult`. If I type `monet`, Emacs puts `has:link\|has:pdf monet` item to the _Selected_ section. If I press `RET` on that item, I get `Wrong type argument: stringp, nil`. If, instead of typing `monet` at the beginning, I simply press `RET`, I get `has:link\|has:pdf` in the _Selected_ section.
> 
> In other words, `bibtex-actions-open` works "normally" only if I remove `has:link\|has:pdf`. But then, why would it be there? The fact that I am asking that question makes me wonder if perhaps I am missing some basic `vertico` concept, or perhaps my Emacs configuration is off?

No, you are not missing anything.

It's an artifact of how `consult-completing-read-multiple` is implemented. Basically, it assumes the prefilter string is a selected candidate, which it's not.

I think I just need to turn that feature off. I will probably do that today.

**Comment by bdarcus** [comment]
BTW, it's possible I'll remove the bibtex-completion dependency, and do this internally.

That would allow a clean slate; so can't hurt to think about how things would ideally work.

Even that exercise can be helpful for writing clear documentation.

**Comment by bdarcus** [comment]
Here's what the `embark-act` menu looks like with changes I committed today.

![Screenshot from 2021-07-27 15-26-53](https://user-images.githubusercontent.com/1134/127215800-c7dc0d3c-98d3-40bf-8c99-8744d654d806.png)

**Comment by salutis** [comment]
I have just upgraded and tried it. This is so much better!

The strange fact is that the associated ePub or JPEG is accessible via "Open PDF", but it is Helm BibTeX's decision to call everything "PDF," from what I understand.

Also, when I press `embark-act` (which I map to the recommended `C-.` key combination), I see the menu twice:

<img width="1187" alt="Screen Shot 2021-07-28 at 10 55 24" src="https://user-images.githubusercontent.com/1158274/127294237-861ce005-654a-47c0-87c5-d9d02b9fbd50.png">

I use the following (recommended) Embark prompter:

```emacs-lisp
(setq-default embark-prompter 'embark-completing-read-prompter)
```

**Comment by bdarcus** [comment]
> I have just upgraded and tried it. This is so much better!
> 
> The strange fact is that the associated ePub or JPEG is accessible via "Open PDF", but it is Helm BibTeX's decision to call everything "PDF," from what I understand.

My point is, we can change that, because it's just a menu description.

https://github.com/bdarcus/bibtex-actions/blob/a2569ef74b40ed6b6459ed1b1b19e12282b426d0/bibtex-actions.el#L175

You can even change it yourself locally.

> Also, when I press `embark-act` (which I map to the recommended `C-.` key combination), I see the menu twice:
> 
> I use the following (recommended) Embark prompter:
> 
> ```emacs-lisp
> (setq-default embark-prompter 'embark-completing-read-prompter)
> ```

There were some changes around this in embark the past few days. Did you update it recently?

**Comment by salutis** [comment]
> WDYT?

So, we have the following terms:

* PDF file -- too specific, somtimes incorrect
* source file -- confusing as my copy is not "the source"
* local file -- a _tiny bit_ too specific, e.g. with a remotely mounted library
* library file -- a bit vague, matches the BibTeX "path" variable

Out of the four candidates, "local" and "library" are perhaps the most precise. Importantly, the UI, the documentation strings, and the code should all use the same terminology.

> There were some changes around this in embark the past few days. Did you update it recently?

Today, I updated with `U` and then `x` in `list-packages`. Every time I do that, I want to see what is new, but there seems to be no such functionality in `package.el` (which I find incredibly surprising, coming from Vim where I could see at least the commits in virtually every package manager).

**Comment by bdarcus** [comment]
Check the embark readme; you may need a config change, like removing that line maybe?

Let's think more about the other stuff. Feel free to post further ideas here.

**Comment by salutis** [comment]
> Check the embark readme; you may need a config change, like removing that line maybe?

The option is still documented in `README.md`. I opened https://github.com/oantolin/embark/issues/301 to find out what is going on over there. (Emacs can be such a Whack-A-Mole.)

> Let's think more about the other stuff.

I love your approach! We all need to think more and do less.

**Comment by bdarcus** [comment]
I'm going to close this; already partially addressed in the new `bibtex-actions-open-library-files`, and working on it in related issues (see, for example, #229).
